File: packages/claude-agent-sdk-adapter/src/hooks/telemetry-hooks.ts
import type {
  TelemetryConfig,
  HookCallback,
  HookCallbackMatcher,
  HookInput,
  HookOutput,
} from "../types";

/**
 * Hook event names from Claude Agent SDK
 */
export type HookEventName =
  | 'SessionStart'
  | 'SessionEnd'
  | 'PreToolUse'
  | 'PostToolUse'
  | 'PostToolUseFailure'
  | 'Stop'
  | 'UserPromptSubmit'
  | 'SubagentStart'
  | 'SubagentStop'
  | 'PreCompact'
  | 'PermissionRequest'
  | 'Notification';

/**
 * Hook configuration for Claude Agent SDK - array of matchers per event
 */
export type HooksConfig = Partial<Record<HookEventName, HookCallbackMatcher[]>>;

/**
 * Telemetry event data for external consumers
 */
export interface TelemetryEvent {
  eventName: string;
  timestamp: number;
  sessionId: string;
  promptName: string;
  data: Record<string, unknown>;
}

/**
 * Telemetry event handler type
 */
export type TelemetryEventHandler = (event: TelemetryEvent) => void | Promise<void>;

/**
 * Creates telemetry hooks for Claude Agent SDK that integrate with AgentMark's
 * tracing system.
 *
 * The hooks capture key events during agent execution:
 * - Session start/end
 * - Tool use (before and after)
 * - Subagent events
 * - Errors
 *
 * @param config - Telemetry configuration
 * @param eventHandler - Optional handler for telemetry events
 * @returns Hook configuration for Claude Agent SDK
 *
 * @example
 * ```typescript
 * const hooks = createTelemetryHooks({
 *   isEnabled: true,
 *   promptName: 'my-agent-task',
 *   props: { userId: '123' },
 * });
 *
 * const result = await query({
 *   prompt: "Do something",
 *   options: { hooks }
 * });
 * ```
 */
export function createTelemetryHooks(
  config: TelemetryConfig,
  eventHandler?: TelemetryEventHandler
): HooksConfig {
  if (!config.isEnabled) {
    return {};
  }

  const emitEvent = async (
    eventName: string,
    sessionId: string,
    data: Record<string, unknown>
  ): Promise<void> => {
    if (eventHandler) {
      await eventHandler({
        eventName,
        timestamp: Date.now(),
        sessionId,
        promptName: config.promptName,
        data: {
          ...data,
          functionId: config.functionId,
          metadata: config.metadata,
          props: config.props,
        },
      });
    }
  };

  const sessionStartHook: HookCallback = async (
    input: HookInput,
    _toolUseId: string | null,
    _options: { signal: AbortSignal }
  ): Promise<HookOutput> => {
    await emitEvent('session_start', input.session_id, {
      cwd: input.cwd,
      transcript_path: input.transcript_path,
    });

    return { continue: true };
  };

  const sessionEndHook: HookCallback = async (
    input: HookInput,
    _toolUseId: string | null,
    _options: { signal: AbortSignal }
  ): Promise<HookOutput> => {
    await emitEvent('session_end', input.session_id, {
      reason: input.reason,
    });

    return { continue: true };
  };

  const preToolUseHook: HookCallback = async (
    input: HookInput,
    toolUseId: string | null,
    _options: { signal: AbortSignal }
  ): Promise<HookOutput> => {
    await emitEvent('tool_start', input.session_id, {
      tool_name: input.tool_name,
      tool_input: input.tool_input,
      tool_use_id: toolUseId,
    });

    return { continue: true };
  };

  const postToolUseHook: HookCallback = async (
    input: HookInput,
    toolUseId: string | null,
    _options: { signal: AbortSignal }
  ): Promise<HookOutput> => {
    await emitEvent('tool_end', input.session_id, {
      tool_name: input.tool_name,
      tool_response: input.tool_response,
      tool_use_id: toolUseId,
    });

    return { continue: true };
  };

  const postToolUseFailureHook: HookCallback = async (
    input: HookInput,
    toolUseId: string | null,
    _options: { signal: AbortSignal }
  ): Promise<HookOutput> => {
    await emitEvent('tool_error', input.session_id, {
      tool_name: input.tool_name,
      error: input.error,
      tool_use_id: toolUseId,
    });

    return { continue: true };
  };

  const stopHook: HookCallback = async (
    input: HookInput,
    _toolUseId: string | null,
    _options: { signal: AbortSignal }
  ): Promise<HookOutput> => {
    await emitEvent('agent_stop', input.session_id, {
      reason: input.reason,
      result: input.result,
    });

    return { continue: true };
  };

  const subagentStartHook: HookCallback = async (
    input: HookInput,
    _toolUseId: string | null,
    _options: { signal: AbortSignal }
  ): Promise<HookOutput> => {
    await emitEvent('subagent_start', input.session_id, {
      subagent_type: input.subagent_type,
      subagent_prompt: input.subagent_prompt,
    });

    return { continue: true };
  };

  const subagentStopHook: HookCallback = async (
    input: HookInput,
    _toolUseId: string | null,
    _options: { signal: AbortSignal }
  ): Promise<HookOutput> => {
    await emitEvent('subagent_stop', input.session_id, {
      subagent_result: input.subagent_result,
    });

    return { continue: true };
  };

  return {
    SessionStart: [{ hooks: [sessionStartHook] }],
    SessionEnd: [{ hooks: [sessionEndHook] }],
    PreToolUse: [{ hooks: [preToolUseHook] }],
    PostToolUse: [{ hooks: [postToolUseHook] }],
    PostToolUseFailure: [{ hooks: [postToolUseFailureHook] }],
    Stop: [{ hooks: [stopHook] }],
    SubagentStart: [{ hooks: [subagentStartHook] }],
    SubagentStop: [{ hooks: [subagentStopHook] }],
  };
}

/**
 * Merges multiple hook configurations together.
 * Each config has arrays of matchers per event - we concatenate the arrays.
 *
 * @param configs - Array of hook configurations to merge
 * @returns Merged hook configuration
 */
export function mergeHooks(...configs: HooksConfig[]): HooksConfig {
  const merged: HooksConfig = {};

  for (const config of configs) {
    for (const [eventName, matchers] of Object.entries(config)) {
      const key = eventName as HookEventName;
      if (merged[key]) {
        merged[key] = [...merged[key]!, ...matchers];
      } else {
        merged[key] = [...matchers];
      }
    }
  }

  return merged;
}

File: packages/claude-agent-sdk-adapter-python/src/agentmark_claude_agent_sdk/hooks/telemetry_hooks.py
"""Telemetry hooks for Claude Agent SDK adapter.

Ported from TypeScript: packages/claude-agent-sdk-adapter/src/hooks/telemetry-hooks.ts
"""

from __future__ import annotations

import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any, Literal, TypedDict

# Hook event names from Claude Agent SDK
HookEventName = Literal[
    "SessionStart",
    "SessionEnd",
    "PreToolUse",
    "PostToolUse",
    "PostToolUseFailure",
    "Stop",
    "UserPromptSubmit",
    "SubagentStart",
    "SubagentStop",
    "PreCompact",
    "PermissionRequest",
    "Notification",
]

# Hook configuration for Claude Agent SDK - dict of event name to list of matchers
HooksConfig = dict[str, list[dict[str, Any]]]


class TelemetryEvent(TypedDict, total=False):
    """Telemetry event data for external consumers."""

    eventName: str
    timestamp: int
    sessionId: str
    promptName: str
    data: dict[str, Any]


# Telemetry event handler type
TelemetryEventHandler = Callable[[TelemetryEvent], None | Awaitable[None]]


@dataclass
class TelemetryConfig:
    """Telemetry configuration for hooks."""

    is_enabled: bool
    prompt_name: str
    props: dict[str, Any]
    function_id: str | None = None
    metadata: dict[str, Any] | None = None


def create_telemetry_hooks(
    config: TelemetryConfig,
    event_handler: TelemetryEventHandler | None = None,
) -> HooksConfig:
    """Create telemetry hooks for Claude Agent SDK that integrate with AgentMark's tracing system.

    The hooks capture key events during agent execution:
    - Session start/end
    - Tool use (before and after)
    - Subagent events
    - Errors

    Args:
        config: Telemetry configuration.
        event_handler: Optional handler for telemetry events.

    Returns:
        Hook configuration for Claude Agent SDK.

    Example:
        hooks = create_telemetry_hooks(TelemetryConfig(
            is_enabled=True,
            prompt_name='my-agent-task',
            props={'userId': '123'},
        ))

        result = await query(
            prompt="Do something",
            options={"hooks": hooks}
        )
    """
    if not config.is_enabled:
        return {}

    async def emit_event(event_name: str, session_id: str, data: dict[str, Any]) -> None:
        if event_handler:
            event: TelemetryEvent = {
                "eventName": event_name,
                "timestamp": int(time.time() * 1000),
                "sessionId": session_id,
                "promptName": config.prompt_name,
                "data": {
                    **data,
                    "functionId": config.function_id,
                    "metadata": config.metadata,
                    "props": config.props,
                },
            }
            result = event_handler(event)
            if result is not None:
                await result

    async def session_start_hook(
        input_data: dict[str, Any],
        _tool_use_id: str | None,
        _options: dict[str, Any],
    ) -> dict[str, Any]:
        await emit_event(
            "session_start",
            input_data.get("session_id", ""),
            {
                "cwd": input_data.get("cwd"),
                "transcript_path": input_data.get("transcript_path"),
            },
        )
        return {"continue": True}

    async def session_end_hook(
        input_data: dict[str, Any],
        _tool_use_id: str | None,
        _options: dict[str, Any],
    ) -> dict[str, Any]:
        await emit_event(
            "session_end",
            input_data.get("session_id", ""),
            {"reason": input_data.get("reason")},
        )
        return {"continue": True}

    async def pre_tool_use_hook(
        input_data: dict[str, Any],
        tool_use_id: str | None,
        _options: dict[str, Any],
    ) -> dict[str, Any]:
        await emit_event(
            "tool_start",
            input_data.get("session_id", ""),
            {
                "tool_name": input_data.get("tool_name"),
                "tool_input": input_data.get("tool_input"),
                "tool_use_id": tool_use_id,
            },
        )
        return {"continue": True}

    async def post_tool_use_hook(
        input_data: dict[str, Any],
        tool_use_id: str | None,
        _options: dict[str, Any],
    ) -> dict[str, Any]:
        await emit_event(
            "tool_end",
            input_data.get("session_id", ""),
            {
                "tool_name": input_data.get("tool_name"),
                "tool_response": input_data.get("tool_response"),
                "tool_use_id": tool_use_id,
            },
        )
        return {"continue": True}

    async def post_tool_use_failure_hook(
        input_data: dict[str, Any],
        tool_use_id: str | None,
        _options: dict[str, Any],
    ) -> dict[str, Any]:
        await emit_event(
            "tool_error",
            input_data.get("session_id", ""),
            {
                "tool_name": input_data.get("tool_name"),
                "error": input_data.get("error"),
                "tool_use_id": tool_use_id,
            },
        )
        return {"continue": True}

    async def stop_hook(
        input_data: dict[str, Any],
        _tool_use_id: str | None,
        _options: dict[str, Any],
    ) -> dict[str, Any]:
        await emit_event(
            "agent_stop",
            input_data.get("session_id", ""),
            {
                "reason": input_data.get("reason"),
                "result": input_data.get("result"),
            },
        )
        return {"continue": True}

    async def subagent_start_hook(
        input_data: dict[str, Any],
        _tool_use_id: str | None,
        _options: dict[str, Any],
    ) -> dict[str, Any]:
        await emit_event(
            "subagent_start",
            input_data.get("session_id", ""),
            {
                "subagent_type": input_data.get("subagent_type"),
                "subagent_prompt": input_data.get("subagent_prompt"),
            },
        )
        return {"continue": True}

    async def subagent_stop_hook(
        input_data: dict[str, Any],
        _tool_use_id: str | None,
        _options: dict[str, Any],
    ) -> dict[str, Any]:
        await emit_event(
            "subagent_stop",
            input_data.get("session_id", ""),
            {"subagent_result": input_data.get("subagent_result")},
        )
        return {"continue": True}

    return {
        "SessionStart": [{"hooks": [session_start_hook]}],
        "SessionEnd": [{"hooks": [session_end_hook]}],
        "PreToolUse": [{"hooks": [pre_tool_use_hook]}],
        "PostToolUse": [{"hooks": [post_tool_use_hook]}],
        "PostToolUseFailure": [{"hooks": [post_tool_use_failure_hook]}],
        "Stop": [{"hooks": [stop_hook]}],
        "SubagentStart": [{"hooks": [subagent_start_hook]}],
        "SubagentStop": [{"hooks": [subagent_stop_hook]}],
    }


def merge_hooks(*configs: HooksConfig) -> HooksConfig:
    """Merge multiple hook configurations together.

    Each config has arrays of matchers per event - we concatenate the arrays.

    Args:
        *configs: Hook configurations to merge.

    Returns:
        Merged hook configuration.
    """
    merged: HooksConfig = {}

    for config in configs:
        if config is None:
            continue
        for event_name, matchers in config.items():
            if event_name in merged:
                merged[event_name] = [*merged[event_name], *matchers]
            else:
                merged[event_name] = [*matchers]

    return merged
